### ISSUE TITLE
NPE in obfuscated code after calling trace.stop()

### ISSUE THREAD

**Issue by sn-michiyo** [firebase/firebase-android-sdk#1383]
### Describe your environment

  * Android Studio version: 3.6.1
  * Firebase Component: Performance
  * Component version: 19.0.0

### Describe the problem

Crashlytics is reporting 80 users experiencing a null pointer exception When calling `trace.stop();`  We haven't seen this during development and trace object itself itsn't null; the crash comes in the obfuscated firebase code.  Is there anything we can to prevent the crash?

```
Fatal Exception: java.lang.NullPointerException
Attempt to invoke virtual method 'com.google.android.gms.internal.firebase-perf.zzcr com.google.firebase.perf.internal.zzq.zzbf()' on a null object reference
com.google.firebase.perf.internal.zzq.zza (zzq.java:39)
com.google.firebase.perf.metrics.zzd.zzcd (zzd.java:20)
com.google.firebase.perf.metrics.Trace.stop (Trace.java:87)
```

app-level build.gradle:
```
implementation 'com.google.firebase:firebase-perf:19.0.0'
...
apply plugin: 'com.google.firebase.firebase-perf'
```

I don't see anything in the [Trace docs](https://developers.google.com/android/reference/com/google/firebase/perf/metrics/Trace#stop()) about NPEs, and didn't find any mention of this in the issues for this repo.

I also looked at the release notes for Performance Monitoring for [19.0.1](https://firebase.google.com/support/release-notes/android#performance_v19-0-1), [19.0.2](https://firebase.google.com/support/release-notes/android#performance_v19-0-2), [19.0.3](https://firebase.google.com/support/release-notes/android#performance_v19-0-3), [19.0.4](https://firebase.google.com/support/release-notes/android#performance_v19-0-4), and [19.0.5](https://firebase.google.com/support/release-notes/android#performance_v19-0-5) for mention of a NPE to see if upgrading would fix the issue, but don't see any mention of it.

**Comment by ramanpreetSinghKhinda** [comment]
Internally Tracking in [b/152218504](http://b/152218504)

**Comment by ramanpreetSinghKhinda** [comment]
Dry run on a Sample App doesn't yield any NPE. Is this reproducible at your end @sn-michiyo?

**Comment by sn-michiyo** [comment]
Right, I haven't been able to reproduce. 

The basic usage looks like 

```
    val firebasePerformance: FirebasePerformance = ...
    val attributes: MutableMap<String, String> = ...
    val newTrace = firebasePerformance.newTrace(name)

    for ((k, v) in attributes) {
        newTrace.putAttribute(k, v)
    }

    newTrace.start()
``` 

and 
```
 val trace: Trace? =  // get trace from data structure where it was stored
 trace?.stop() ?: return
```

We are not using Trace.CREATOR.  There may be multiple traces that have been started and not yet stopped.  

I belatedly realized that the crash has affected many more users if I expand the time period, so I just looked at the first 100 crash reports.  While the "stop trace" code is being called from four different places in our app, the stack traces (that I looked at) only come from two of them.  

The crashing call to the "stop tracking" code originates in a call to Completable's `doFinally` block in one case or in a call to Single's `doFinally` block in the other.  The other two usages of "stop trace" code, which I did not see in the stack traces, do not originate from inside rxjava code. 

In case it's relevant: 
```
    implementation 'io.reactivex.rxjava2:rxandroid:2.1.1'
    implementation 'io.reactivex.rxjava2:rxjava:2.2.0'
```

**Comment by sn-michiyo** [comment]
I should note, though, that the calls to "stop tracking" code in the `doFinally` blocks happen a lot more frequently than the other calls to "stop tracking" code.

**Comment by sn-michiyo** [comment]
@ramanpreetSinghKhinda would putting the call to `stop()` inside a `try` block have any negative repercussions?  (Beyond not getting results for that particular trace, but presumably we aren't getting those anyway when the app crashes.)

**Comment by ramanpreetSinghKhinda** [comment]
Thanks @sn-michiyo for the details.

I am preparing a fix for this and plan to rollout in the upcoming release. Will update here with more details on that.

**Comment by ramanpreetSinghKhinda** [comment]
Regarding [this](https://github.com/firebase/firebase-android-sdk/issues/1383#issuecomment-605059129) comment: `There should not be any negative effect due to this`

**Comment by ramanpreetSinghKhinda** [comment]
Hi @sn-michiyo ,

I have some more questions:

Q1. Can you explain what type of data structure are you using to `store` and thus `retrieve` the trace object?

Q2. Are you passing the `trace` object within activities? For e.g. using Intent or something?

Q3. Are you accessing the `trace` object outside of the lifecycle of the activity? For e.g is the action `closing the app and then reopening` trying to access your previously created trace objects from your data structure?

**Comment by MichalFrSN** [comment]
Hi @ramanpreetSinghKhinda , Following your conversation with @sn-michiyo:
We store the Trace object locally on ApplicationScope class within a MutableMap:
`val customTraceMap : MutableMap<String, Trace> `

We don't pass the Trace obj within activities, we create & store the `Trace` object in the same class:
`        for ((k, v) in mutableAttributes) {
            newTrace.putAttribute(k, v)
        }
`
`
        customTraceMap[name] = newTrace
        customTraceMap[name]?.start()
`

and stop the traces:
`            customTraceMap[name]?.stop() ?: return
`

**Comment by ramanpreetSinghKhinda** [comment]
This should be fixed in `v19.0.6`. 

[Release Notes](https://firebase.google.com/support/release-notes/android#performance_v19-0-6)

Please reopen if the issue persists.
